### PR TITLE
Optimize LayernormGrad

### PR DESCRIPTION
**Pull request by weixingzhang**
Previously, both the input(X) and the output(Y) of Layernorm are stashed for computation of LayernormGrad. Actually, we can just stash Y and then compute X = (Y-Bias)/Scale during LayernormGrad computation.

With this change, the memory consumption during training time will be reduced and the batch size can be increased.

  | Baseline | Invertible Layernorm Grad
-- | -- | --
Phase 1 Max BS | 66 | 72
Phase 1 Throughput   (ex/sec) | 185.534 | 186.816
Phase 2 Max BS | 10 | 11
Phase 2 Throughput   (ex/sec) | 36.0552 | 35.6761

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/layer_norm.cc [hunk 1]
```diff
@@ -24,7 +24,7 @@ namespace cuda {
       LayerNorm<T, U>);
 
 REGISTER_KERNEL_TYPED(float, float)
-REGISTER_KERNEL_TYPED(double, float)
+REGISTER_KERNEL_TYPED(double, double)
```
Comment (yuslepukhin) [inline onnxruntime/contrib_ops/cuda/layer_norm.cc:27]: >D(double, double) [](start = 20, length = 17)

Would it possible to combine these 3 in a single declaration? You can use BuildKernelDefConstraints() to combine so less typing.

**Comment by yuslepukhin** [comment]
@snnn is added to the review. #Closed

**Comment by yuslepukhin** [comment]
>       .SinceVersion(1)

This changes types. I think there was a conversation to change versioning on contrib ops as well. @chasun@microsoft.com may remember

---
Refers to: orttraining/orttraining/core/graph/gradient_schema_defs.cc:1345 in 832fca2. [](commit_id = 832fca23d920b49f7a332492462ab3b944dedcfb, deletion_comment = False)

**Comment by yuslepukhin** [comment]
>   ORT_ENFORCE(M != 1);

We want more in the message here to help diagnostics.

---
Refers to: orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:49 in 832fca2. [](commit_id = 832fca23d920b49f7a332492462ab3b944dedcfb, deletion_comment = False)

File: orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc [hunk 2]
```diff
@@ -36,12 +36,13 @@ LayerNormGrad<T>::LayerNormGrad(const OpKernelInfo& op_kernel_info)
...
-  const Tensor* X = op_kernel_context->Input<Tensor>(1);
+  const Tensor* Y = op_kernel_context->Input<Tensor>(1);
   const Tensor* scale = op_kernel_context->Input<Tensor>(2);
-  const Tensor* mean = op_kernel_context->Input<Tensor>(3);
+  const Tensor* bias = op_kernel_context->Input<Tensor>(3);
   const Tensor* inv_std_var = op_kernel_context->Input<Tensor>(4);
 
-  const auto& X_shape = X->Shape();
+  const auto& Y_shape = Y_grad->Shape();
+  const auto& X_shape = Y_shape;
```
Comment (yuslepukhin) [inline orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:45]: Is this not the same thing?

File: orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc [hunk 3]
```diff
@@ -54,9 +55,9 @@ Status LayerNormGrad<T>::Compute(OpKernelContext* op_kernel_context) const {
 
   // Note: Eigen has column-major storage order by default
   ConstEigenArrayMap<T> Y_grad_arr{Y_grad->Data<T>(), M, N};
-  ConstEigenArrayMap<T> X_arr{X->Data<T>(), M, N};
+  ConstEigenArrayMap<T> Y_arr{Y->Data<T>(), M, N};
   ConstEigenVectorArrayMap<T> scale_vec{scale->Data<T>(), M};
-  ConstEigenVectorArrayMap<float> mean_vec{mean->Data<float>(), N};
+  ConstEigenVectorArrayMap<T> bias_vec{bias->Data<T>(), M};
   ConstEigenVectorArrayMap<float> inv_std_var_vec{inv_std_var->Data<float>(), N};
```
Comment (yuslepukhin) [inline orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:61]: >inv_std_var [](start = 50, length = 11)

Is this a training only data/kernel? Is this available during inference?

**Comment by tianleiwu** [comment]
It is a kernel also used in inference (like GPT-2). Please avoid compute gradient when the optional output is not needed.

File: orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc (hunk 3 above)
Comment (tianleiwu) [reply orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:61]: There is a LayerNormlization kernel:
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/contrib_ops/cpu/layer_norm.cc
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/contrib_ops/cuda/layer_norm.cc

That is also used in GPT-2 inference. Please take some time to avoid gradient computing for inference (some outputs are optional, so do not compute them when they do not exist in node output).

Comment (weixingzhang) [reply orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:61]: This implementation is only for training now. Inference has its own implementation.

File: orttraining/orttraining/training_ops/cuda/nn/layer_norm.cc [hunk 4]
```diff
@@ -35,18 +35,19 @@ Status LayerNormGrad<T, U>::ComputeInternal(OpKernelContext* p_op_kernel_context
...
-  auto X_data = reinterpret_cast<const CudaT*>(X->template Data<T>());
+  auto Y_data = reinterpret_cast<const CudaT*>(Y->template Data<T>());
   auto scale_data = reinterpret_cast<const CudaT*>(scale->template Data<T>());
-  auto mean_data = reinterpret_cast<const CudaU*>(mean->template Data<U>());
+  auto bias_data = reinterpret_cast<const CudaT*>(bias->template Data<T>());
   auto inv_std_var_data = reinterpret_cast<const CudaU*>(inv_std_var->template Data<U>());
 
-  const TensorShape& x_shape = X->Shape();
+  const TensorShape& y_shape = Y->Shape();
+  const TensorShape& x_shape = y_shape;
```
Comment (ytaous) [inline orttraining/orttraining/training_ops/cuda/nn/layer_norm.cc:50]: nit: inconsistent variable names vs. the other ones you used in training_ops/cpu/nn/layer_norm,cc

File: onnxruntime/contrib_ops/cuda/layer_norm.cc (hunk 1 above)
Comment (weixingzhang) [reply onnxruntime/contrib_ops/cuda/layer_norm.cc:27]: The kernel definition will be non-typed. But we need to handle type in the kernel implementation. right? If so, we need to add more code. It seems it is better to have a separate PR for this kind of change.

File: orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc (hunk 2 above)
Comment (weixingzhang) [reply orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:45]: In the new op schema, there is no X as input.

File: orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc (hunk 3 above)
Comment (weixingzhang) [reply orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:142]: @tianleiwu this PR is to optimize LayernormGrad. Regarding to Layernorm, we can create another work item for it.

Comment (weixingzhang) [reply orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:142]: @yuslepukhin this is training only kernel.

File: orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu [hunk 5]
```diff
@@ -475,9 +498,17 @@ __global__ void cuComputePartGradGammaBeta(
   U* warp_buf2 = warp_buf1 + blockDim.y * blockDim.y * row_stride;
   // compute partial sums from strided inputs
   // do this to increase number of loads in flight
-  cuLoadWriteStridedInputs(i1_beg, thr_load_row_off, thr_load_col_off, i2_off, row_stride, warp_buf1, warp_buf2, input, dout, i1_end, n2, mean, invvar);
+  if (use_mean) {
+    cuLoadWriteStridedInputs<T, U, use_mean>(i1_beg, thr_load_row_off, thr_load_col_off, i2_off, row_stride, warp_buf1, warp_buf2, input, output, dout, i1_end, n2, gamma, beta, mean, invvar);
```
Comment (ashbhandare) [inline orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu:502]: is the if-else construct intentional even when the function calls look identical?

File: orttraining/orttraining/models/bert/main.cc [hunk 6]
```diff
@@ -452,12 +454,15 @@ Status ParseArguments(int argc, char* argv[], BertParameters& params, OrtParamet
         "Log severity must be in the range [", static_cast<int>(logging::Severity::kVERBOSE),
         ", ", static_cast<int>(logging::Severity::kFATAL), "].");
     ort_params.vlog_level = flags["ort_vlog_level"].as<int>();
+
+    params.use_invertible_layernorm = flags["use_invertible_layernorm"].as<bool>();
```
Comment (ashbhandare) [inline orttraining/orttraining/models/bert/main.cc:458]: probably need to add a way to specify this using the pytorch frontend too

File: orttraining/orttraining/test/training_ops/cuda/layer_norm_test.cc [hunk 7]
```diff
@@ -97,5 +97,95 @@ TEST(CudaKernelTest, LayerNormGrad_LargeSizeTensor) {
...
+
+  const auto N = std::accumulate(n_dims.begin(), n_dims.end(), static_cast<int64_t>(1), std::multiplies<>{});
+  const auto M = std::accumulate(m_dims.begin(), m_dims.end(), static_cast<int64_t>(1), std::multiplies<>{});
+
+  CompareOpTester test{"InvertibleLayerNormalizationGrad", 1, kMSDomain};
+
+  test.AddAttribute("axis", axis);
+
+  RandomValueGenerator random{};
+  const auto Y_grad_data = random.Uniform<float>(n_x_m_dims, k_random_data_min, k_random_data_max);
```
Comment (ashbhandare) [inline orttraining/orttraining/test/training_ops/cuda/layer_norm_test.cc:116]: I didn't see any mixedprecision tests, might need to add them

File: orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu (hunk 5 above)
Comment (weixingzhang) [reply orttraining/orttraining/training_ops/cuda/nn/layer_norm_impl.cu:502]: good catch. Yeah, it is not necessary to have both.

File: orttraining/orttraining/test/training_ops/cuda/layer_norm_test.cc (hunk 7 above)
Comment (weixingzhang) [reply orttraining/orttraining/test/training_ops/cuda/layer_norm_test.cc:116]: Agree. But current test framework such as implementation of CompareWithCPU needs to be changed to support fp16. It is better to have a separate PR for it.

File: orttraining/orttraining/models/bert/main.cc (hunk 6 above)
Comment (weixingzhang) [reply orttraining/orttraining/models/bert/main.cc:458]: yes, will add it.
